### ISSUE TITLE
EOFError in transport.py when connecting to Fortinet Device

### ISSUE THREAD

**Issue by adieball** [paramiko/paramiko#687]
Hi Guys

----
Paramiko: 1.16.0
Python: 2.7.10
OS: Mac OS 10.11.3
----

I have a vey simple script trying to log into a Fortinet (Fortigate) device:

```
import paramiko

if __name__ == "__main__":
    ip = "10.x.x.x."
    username = "user"
    password = "password"
    
    import logging
    logging.basicConfig()
    logging.getLogger('paramiko.transport').setLevel(logging.DEBUG)

    ssh = paramiko.SSHClient()
    ssh.set_missing_host_key_policy(paramiko.AutoAddPolicy())
    ssh.connect(ip,username=username,password=password)
    ssh_transport = ssh.get_transport()

    for command in ("get system status"):
        chan = ssh_transport.open_session()
        chan.exec_command(command)
        exit_code = chan.recv_exit_status()
        stdin = chan.makefile('wb', -1)         # pylint: disable-msg=W0612
        stdout = chan.makefile('rb', -1)
        stderr = chan.makefile_stderr('rb', -1)  # pylint: disable-msg=W0612
        output = stdout.read()
        print output
```

Whatever I do, when I try to connect (no even getting to the authentication), I get a traceback:

> DEBUG:paramiko.transport:starting thread (client mode): 0x94122d0L
> DEBUG:paramiko.transport:Local version/idstring: SSH-2.0-paramiko_1.16.0
> DEBUG:paramiko.transport:Remote version/idstring: SSH-2.0-a3WzN
> INFO:paramiko.transport:Connected (version 2.0, client a3WzN)
> DEBUG:paramiko.transport:kex algos:[u'diffie-hellman-group-exchange-sha1', u'diffie-hellman-group1-sha1'] server key:[u'ssh-rsa', u'ssh-dss'] client encrypt:[u'aes128-cbc', u'3des-cbc', u'blowfish-cbc', u'cast128-cbc', u'arcfour', u'aes192-cbc', u'aes256-cbc', u'rijndael-cbc@lysator.liu.se', u'aes128-ctr', u'aes192-ctr', u'aes256-ctr'] server encrypt:[u'aes128-cbc', u'3des-cbc', u'blowfish-cbc', u'cast128-cbc', u'arcfour', u'aes192-cbc', u'aes256-cbc', u'rijndael-cbc@lysator.liu.se', u'aes128-ctr', u'aes192-ctr', u'aes256-ctr'] client mac:[u'hmac-md5', u'hmac-sha1', u'hmac-ripemd160', u'hmac-ripemd160@openssh.com', u'hmac-sha1-96', u'hmac-md5-96'] server mac:[u'hmac-md5', u'hmac-sha1', u'hmac-ripemd160', u'hmac-ripemd160@openssh.com', u'hmac-sha1-96', u'hmac-md5-96'] client compress:[u'none', u'zlib'] server compress:[u'none', u'zlib'] client lang:[u''] server lang:[u''] kex follows?False
> DEBUG:paramiko.transport:Kex agreed: diffie-hellman-group1-sha1
> DEBUG:paramiko.transport:Cipher agreed: aes128-ctr
> DEBUG:paramiko.transport:MAC agreed: hmac-md5
> DEBUG:paramiko.transport:Compression agreed: none
> DEBUG:paramiko.transport:EOF in transport thread
> Traceback (most recent call last):
>   File "/Users/username/workspace/FortiNet/src/test.py", line 17, in <module>
>     ssh.connect(ip,username=username,password=password)
>   File "build/bdist.macosx-10.11-intel/egg/paramiko/client.py", line 325, in connect
>   File "build/bdist.macosx-10.11-intel/egg/paramiko/transport.py", line 492, in start_client
> EOFError

I did some further research and it looks that it works fine when the Fortigate is running Code 5.2.4 or below, but causing issues when running 5.2.5 upwards.

Attached are also two tcp dumps showing the connection point of view. One boring with 5.2.4 one not working with 5.2.5 or 5.4.0

Although I understand that you cannot support each and every device, I think it might be useful to understand why it's not working. Please let me know if I can do anything to provide you with more details on this error (quite a newbie in python).

I know there was a vulnerability problem with those devices and a python script using paramiko to use this exploit, so I also (as a matter of curiosity) changed the Client ID from "paramiko" to OpenSSH_6.4, just to be sure there is not some internal code refusing any paramiko connection, but I still got to the exact same error.

thanks

[Archive.zip](https://github.com/paramiko/paramiko/files/128393/Archive.zip)

**Comment by luk206** [comment]
I have the same problem with paramiko, and with one golang app too, it doesn't matter the server from you try the ssh connection.

**Comment by adieball** [comment]
with a tip from a user on Stackoverflow I was able to fix this. 

config system global
   - set dh-parms 1024

This sets the minimum expected DH to 1024 instead of 3072 which is the default. This is kinda weird, as according to the docs, this affects the SSL bit in the fortigate only, no word about ssh.
I'll also ask Fortinet about this in parallel and report here.

**Comment by luk206** [comment]
In fortiOS 5.2 there is not the "set dh-parms 1024" option.

The most similar is:
on: config firewall ssl setting
    set ssl-dh-bits 1024

but, unfortunately I still can't connect by ssh and paramiko, or golang. Thanks anyway.

**Comment by adieball** [comment]
Correct, I realised this as well over the weekend. I asked Fortinet about this

**Comment by luk206** [comment]
We have one open issue with fortigate too, but still without anwer. God luck to everybody...

**Comment by rcarrillo** [comment]
Fortigate announces `diffie-hellman-group1-sha1` as an acceptable parameter, but when a client tries to use it, it rejects the connection.
Despite Paramiko having the [weak](https://weakdh.org/) `diffie-hellman-group1-sha1` as the first preferred key exchange algorithm, the Fortigate device must not says it accepts a parameter it doesn't. So, this is a Fortigate bug.
If the client uses the other acceptable key exchange `diffie-hellman-group-exchange-sha1`, the connection is successful, as you can see with:
```sh
ssh -vv -oKexAlgorithms=diffie-hellman-group1-sha1 fortigate
# connection closed by fortigate

ssh -vv -oKexAlgorithms=diffie-hellman-group-exchange-sha1 fortigate
# connection successful
```
**Workaround**
Edit the preferred key exchange algorithms in `transport.py` https://github.com/paramiko/paramiko/blob/74ba0149347bfeb2f83ddd46672a2912aea51f23/paramiko/transport.py#L125-L130
Either move `diffie-hellman-group1-sha1` bellow `diffie-hellman-group-exchange-sha1` or just delete `diffie-hellman-group1-sha1` from the tuple.
I think `paramiko.transport.SecurityOptions` could be used too, but I didn't find a trivial way to do it. I'll try again later.

**Comment by adieball** [comment]
this makes totally sense.
I forwarded your information to Fortinet Development. Will keep you posted here when I hear back from them.

Andre

**Comment by shanemadden** [comment]
@adieball Have you heard back from Fortinet's support on this?  I'm guessing their fix for your issue was the new kex packet that's being sent in 5.2.7 that offers only `diffie-hellman-group-exchange-sha1` and nothing else, which I assume is from this bug id in the release notes:

> 300588 - Cannot connect SSH to FGT with kex algorithm order dh-group1-sha1, dh-group14-sha1 and dh-group-exchange-sha1.

I guess I'm putting a new ticket in to them on this, since my problem is that the client library I'm using doesn't support dh-group-exchange-sha1 at all, and it's now all they offer (with no config flags to loosen the behavior).

**Comment by adieball** [comment]
thanks @shanemadden I haven't seen this yet.
Keep me posted on the result, but I doubt they'll do anything more within the near future. Maybe you can update your client?

**Comment by bitprophet** [comment]
From a quick skim the only thing I can see that Paramiko could do to help with this issue is to make modifying the preferred algorithms behavior, easier - that sound accurate? I feel like there's open tickets around that but haven't dug any up yet.

**Comment by jworl** [comment]
Just wanted to confirm that rcarillo's workaround is successful for me. Moving diffie-hellman-group1-sha-1 below the diffie-hellman-group-exchange-* entries fixes the Paramiko error we received. Thank you so much!
